tactics/class_tactics.ml

```diff
@@ -236,16 +236,22 @@ let e_give_exact flags poly (c,clenv) =
   Sigma.Unsafe.of_pair (Clenvtac.unify ~flags t1 <*> exact_no_check c, sigma)
   end }
 
+let clenv_unique_resolver_tac with_evars ~flags clenv' =
+  Proofview.Goal.enter { enter = begin fun gls ->
+    let resolve =
+      try Proofview.tclUNIT (clenv_unique_resolver ~flags clenv' gls)
+      with e -> Proofview.tclZERO e
+    in resolve >>= fun clenv' ->
+       Clenvtac.clenv_refine with_evars ~with_classes:false clenv'
+  end }
+
 let unify_e_resolve poly flags = { enter = begin fun gls (c,_,clenv) ->
   let clenv', c = connect_hint_clenv poly c clenv gls in
-  let clenv' = clenv_unique_resolver ~flags clenv' gls in
-    Clenvtac.clenv_refine true ~with_classes:false clenv'
-  end }
+  clenv_unique_resolver_tac true ~flags clenv' end }
 
 let unify_resolve poly flags = { enter = begin fun gls (c,_,clenv) ->
   let clenv', _ = connect_hint_clenv poly c clenv gls in
-  let clenv' = clenv_unique_resolver ~flags clenv' gls in
-    Clenvtac.clenv_refine false ~with_classes:false clenv'
+  clenv_unique_resolver_tac false ~flags clenv'
```

Comment:
@cmangin Indeed, and this is what I had done in 9ab3d55. Feel free to cherry-pick but I've dropped it from my own PR.